// @ims/gateway-hub/Lib/CommType/ssl.js
/* ************************************************************************** */
const tls            = require('tls');
const errorModule    = require('../../Utils/error');
const utils          = require('../../Utils/timestamp');
const parseUrl 					= require('url').parse;
const fs 				= require('fs');
const mkdirp 				= require('mkdirp');

/* ************************************************************************** */


/* ************************************************************************** */
module.exports = function talkToServer (transXpedia) {
	return new Promise ((resolved, rejected) => {
		const url = parseUrl(transXpedia.clientData.gatewayUrl);
		let currentEventGen = undefined;
		transXpedia.responseData.gatewayConnTime  = utils.getTimeStamp();
		let tlsClient = tls.connect(url.port, url.hostname, {
			rejectUnauthorized : transXpedia.rejectUnauthorized ? transXpedia.rejectUnauthorized : 'false',
			ca : transXpedia.clientData.gatewayCaCertificate,

		},function() {
			transXpedia.responseData.gatewaySendTime  = utils.getTimeStamp();
			if(transXpedia.clientData.debugFlag == true){
				let backupPath = './DebugDumps';
				let filename = `dataTOGateway_${transXpedia.clientData.rrn||transXpedia.clientData.transactionId}`;

				mkdirp(backupPath)
						.then(()=>{
							fs.writeFile(`${backupPath}/${filename}`, Buffer.from (transXpedia.requestBuffer.buffer), (e) => {
								if(e)
										console.log("error in creating dataTOGateway File",e);
								else
										console.log("dataTOGateway File is created");
						});
						})
						.catch((e)=>{
							console.log("error in creating directory ./DebugDumps",e);
						})
				}
			tlsClient.write(Buffer.from (transXpedia.requestBuffer.buffer));

		});

		tlsClient.on('data', recieveData).on('error', connectionError)
			.on('close', connectionClose).on('timeout', connectionTimeOut)
			.on('connect', connection);


		tlsClient.setTimeout (transXpedia.clientData.gatewayTimeout);
		function recieveData (RMessageBuffer) {
			currentEventGen = 'data';
			tlsClient.destroy();
			transXpedia.CLIENT_JSON.GATEWAY_RESPBACK_TIME  = utils.getTimeStamp();
			if(transXpedia.clientData.debugFlag == true){
				let backupPath = './DebugDumps';
				let filename = `FromGatewayToFI_${transXpedia.clientData.rrn || transXpedia.clientData.transactionId}`;
				mkdirp(backupPath)
				.then(()=>{
					fs.writeFile(`${backupPath}/${filename}`, Buffer.from (RMessageBuffer), (e) => {
						if(e)
								console.log("error in creating fromGatewayToFI File",e);
						else
								console.log("fromGatewayToFI File is created");
				});
				})
				.catch((e)=>{
					console.log("error in creating directory ./DebugDumps",e);
				})
	
				}
			transXpedia.RESPONSE_BUFFER = RMessageBuffer;
			return resolved(transXpedia);

		}

		function connectionError (e) {
			currentEventGen = 'error';
			transXpedia.responseData.gatewayRespBackTime  = utils.getTimeStamp();
			transXpedia.errorData.systemErr 	= e;
			transXpedia.errorData.serverError = errorModule.E_GATEWAY_CONNECTION;
			transXpedia.errorData.errDescription = `GATEWAY CONNECTION FAILURE` +
				`[FUNCTION : talkToServer] ` +
				`[RRN : ${transXpedia.clientData.rrn}] ` +
				`[TRANSACTION_ID : ${transXpedia.clientData.transactionId}]`;
			return rejected(transXpedia);

		}

		function connectionClose (hadError) {
			transXpedia.responseData.gatewayRespBackTime  = utils.getTimeStamp();
			if (hadError == true) {
				transXpedia.errorData.systemErr 	= 'hadError : ' + hadError;
				transXpedia.errorData.serverError = errorModule.E_GATEWAY_TRANSMISSION;
				transXpedia.errorData.errDescription = `GATEWAY TRANSMISSION ERROR ` +
					`[FUNCTION : talkToServer] ` +
					`[RRN : ${transXpedia.clientData.rrn}] ` +
					`[TRANSACTION_ID : ${transXpedia.clientData.transactionId}]`;
				return rejected(transXpedia);
			} else if (undefined == currentEventGen) {
				transXpedia.errorData.systemErr 	= 'NA';
				transXpedia.errorData.serverError = errorModule.E_GATEWAY_NO_DATA_IN_RESP;
				transXpedia.errorData.errDescription = `NO DATA RECIEVED FROM GATEWAY ` +
				`[FUNCTION : talkToServer] ` +
				`[RRN : ${transXpedia.clientData.rrn}] ` +
				`[TRANSACTION_ID : ${transXpedia.clientData.transactionId}]`;
				return rejected(transXpedia);
			} 

		}

		function connectionTimeOut () {
			currentEventGen = 'timeout';
			tlsClient.destroy();
			transXpedia.responseData.gatewayRespBackTime  = utils.getTimeStamp();
			transXpedia.errorData.systemErr 	= null;
			transXpedia.errorData.serverError = errorModule.E_GATEWAY_TIMEDOUT;
			transXpedia.errorData.errDescription = `GATEWAY TIMEDOUT ` +
			`[FUNCTION : talkToServer] ` +
			`[RRN : ${transXpedia.clientData.rrn}] ` +
			`[TRANSACTION_ID : ${transXpedia.clientData.transactionId}]`;
			return rejected(transXpedia);
		}

		function connection () {
		}

	});

};
/* ************************************************************************** */
