// @ims/gateway-hub/Gateways/MMI/ajvSchemas.js
const getAuthTokenSchema = {
  type: "object",
  title: "GET AUTH TOKEN",
  properties: {
    gatewayUrl: { type: "string",minLength:1,pattern: "https?" },
    gatewayTimeout: { type: "number"},
    transactionId: { type: "string",minLength:1 },
    clientId: { type: "string", minLength:1 },
    clientSecret: { type: "string", minLength:1 },
    channelId: { type: "string",minLength:1 },
    userId: { type: "string",minLength:1 },
    bankCode: { type: "string",minLength:1 },
    transactionBranch: { type: "string",minLength:1 },
    gatewayCaCertificate : { type: "string" },
    key: { type: "string" },
    cert: { type: "string" },
  },
  required: [
    "gatewayUrl",
    "gatewayTimeout",
    "transactionId",
    "clientId",
    "clientSecret",
    "channelId",
    "userId",
    "bankCode",
    "transactionBranch"
  ],
};

const getRevGeoCodeSchema = {
  type: "object",
  title: "GET REV GEOCODE",
  properties: {
    gatewayUrl: { type: "string",minLength:1,pattern: "https?" },
    gatewayTimeout: { type: "number"},
    transactionId: { type: "string",minLength:1 },
    clientId: { type: "string", minLength:1 },
    clientSecret: { type: "string", minLength:1 },
    channelId: { type: "string",minLength:1 },
    userId: { type: "string",minLength:1 },
    bankCode: { type: "string",minLength:1 },
    transactionBranch: { type: "string",minLength:1 },
    accessToken : { type: "string", minLength:1},
    latitude : { type: "string", minLength:1},
    longitude : { type: "string", minLength:1},
    gatewayCaCertificate : { type: "string" },
    key: { type: "string" },
    cert: { type: "string" },
  },
  required: [
    "gatewayUrl",
    "gatewayTimeout",
    "transactionId",
    "clientId",
    "clientSecret",
    "channelId",
    "userId",
    "bankCode",
    "transactionBranch",
    "accessToken",
    "latitude",
    "longitude"
  ],
};


module.exports = {
  getAuthTokenSchema,
  getRevGeoCodeSchema,
}
