// @ims/gateway-hub/Gateways/APS/ajvSchemas.js
const creditCardPreInsurance = {
  type: "object",
  title: "GET AUTH TOKEN",
  properties: {
    requestNumber: { type: "string", minLength: 12, maxLength: 12 },
    applicationRefNo: { type: "string", minLength: 16, maxLength: 16 },
    base64encodedDocument: { type: "string", minLength: 1 },
    advReferenceNumber: { type: "string", minLength: 1 },
    rrn: { type: "string", minLength: 12, maxLength: 12 },
    consentDate: { type: "string" }, //YYYYMMDDHHMMSS
    consentFlag: { type: "string", enum: ["Y", "N"] },
    consentLanguage: { type: "string", minLength: 5 },
    consentVersion: { type: "string", minLength: 1 },
    gatewayUrl: { type: "string", minLength: 1, pattern: "https?" },
    gatewayTimeout: { type: "number" },
    gatewayCaCertificate: { type: "string" },
    key: { type: "string" },
    cert: { type: "string" },
  },
  required: [
    "requestNumber",
    "applicationRefNo",
    "base64encodedDocument",
    "advReferenceNumber",
    "rrn",
    "consentDate",
    "consentFlag",
    "consentLanguage",
    "consentVersion",
    "gatewayUrl",
    "gatewayTimeout",
  ],
};

module.exports = {
  creditCardPreInsurance,
};
