// @ims/gateway-hub/Gateways/CBI/ajvSchemas.js
const ACCTandCIFCreationSchema = {
  type: "object",
  title: "ACCOUNT AND CIF CREATION",
  properties: {
    aofno: {
      type: "string",
    },
    enrollmenttype: {
      type: "string",
    },
    accsubtype: {
      type: "string",
    },
    branch: {
      type: "string",
    },
    custtittle: {
      type: "string",
    },
    custname: {
      type: "string",
    },
    custmiddlename: {
      type: "string",
    },
    custlastname: {
      type: "string",
    },
    custdob: {
      type: "string",
    },
    custage: {
      type: "string",
    },
    custsex: {
      type: "string",
    },
    cusrmartialstatus: {
      type: "string",
    },
    mothertitle: {
      type: "string",
    },
    motherfirstname: {
      type: "string",
    },
    motherlastname: {
      type: "string",
    },
    custcommunitycode: {
      type: "string",
    },
    custnationality: {
      type: "string",
    },
    mobilecountrycode: {
      type: "string",
    },
    custaadhaarno: {
      type: "string",
    },
    custmobilenumber: {
      type: "string",
    },
    custemailid: {
      type: "string",
    },
    custplaceofbirth: {
      type: "string",
    },
    educationalqualification: {
      type: "string",
    },
    presentaddresstype: {
      type: "string",
    },
    custaddress1: {
      type: "string",
    },
    custaddress2: {
      type: "string",
    },
    custaddress3: {
      type: "string",
    },
    custpresentlandmark: {
      type: "string",
    },
    custpresentlocality: {
      type: "string",
    },
    custvillage: {
      type: "string",
    },
    custcitycode: {
      type: "string",
    },
    custsubdist: {
      type: "string",
    },
    custdistrict: {
      type: "string",
    },
    custstatecode: {
      type: "string",
    },
    custcountry: {
      type: "string",
    },
    custpinnumber: {
      type: "string",
    },
    custpresentpostoffice: {
      type: "string",
    },
    permaddresstype: {
      type: "string",
    },
    custpermaddress1: {
      type: "string",
    },
    custpermaddress2: {
      type: "string",
    },
    custpermaddress3: {
      type: "string",
    },
    custpermlandmark: {
      type: "string",
    },
    custpermlocality: {
      type: "string",
    },
    custpermvillage: {
      type: "string",
    },
    custpermcitycode: {
      type: "string",
    },
    custpermsubdist: {
      type: "string",
    },
    custpermdistrict: {
      type: "string",
    },
    custpermstatecode: {
      type: "string",
    },
    permcountry: {
      type: "string",
    },
    custpermpincode: {
      type: "string",
    },
    custpermpostoffce: {
      type: "string",
    },
    custresidenttype: {
      type: "string",
    },
    custresidenceperiod: {
      type: "string",
    },
    custnetworth: {
      type: "string",
    },
    custocupationcode: {
      type: "string",
    },
    customerevaluationrequired: {
      type: "string",
    },
    custpancardpresense: {
      type: "string",
    },
    custtype: {
      type: "string",
    },
    custhousetype: {
      type: "string",
    },
    custemployertype: {
      type: "string",
    },
    custincomesource: {
      type: "string",
    },
    custtaxslab: {
      type: "string",
    },
    custform60or61: {
      type: "string",
    },
    custformpancardnumber: {
      type: "string",
    },
    custvehicletype: {
      type: "string",
    },
    custlanguagecode: {
      type: "string",
    },
    custannualcredit: {
      type: "string",
    },
    annualincome: {
      type: "string",
    },
    addressproofdocnumber: {
      type: "string",
    },
    addressproofdoctype: {
      type: "string",
    },
    addrissuedate: {
      type: "string",
    },
    addrissuedby: {
      type: "string",
    },
    addrissueplace: {
      type: "string",
    },
    identitydocnumber: {
      type: "string",
    },
    identitydoctype: {
      type: "string",
    },
    idexpirydate: {
      type: "string",
    },
    idissuedate: {
      type: "string",
    },
    idissuedby: {
      type: "string",
    },
    idissueplace: {
      type: "string",
    },
    relationshiptype: {
      type: "string",
    },
    relationshipname: {
      type: "string",
    },
    relationshiptitle: {
      type: "string",
    },
    relfirstname: {
      type: "string",
    },
    rellastname: {
      type: "string",
    },
    nomineetitlecode: {
      type: "string",
    },
    nomineename: {
      type: "string",
    },
    nomineelastname: {
      type: "string",
    },
    nomineedob: {
      type: "string",
    },
    nomineeage: {
      type: "string",
    },
    nomineeaddress: {
      type: "string",
    },
    nomineerelation: {
      type: "string",
    },
    nomineetelephone: {
      type: "string",
    },
    nomineeminor: {
      type: "string",
    },
    guardianname: {
      type: "string",
    },
    guardianaddress1: {
      type: "string",
    },
    housingloan: {
      type: "string",
    },
    carloan: {
      type: "string",
    },
    educationalloan: {
      type: "string",
    },
    consumerloan: {
      type: "string",
    },
    businessagricultureloan: {
      type: "string",
    },
    ownbankfixeddeposit: {
      type: "string",
    },
    owngold: {
      type: "string",
    },
    ownmutualfund: {
      type: "string",
    },
    ownnsckvc: {
      type: "string",
    },
    ownppf: {
      type: "string",
    },
    ownproperty: {
      type: "string",
    },
    ownrbibond: {
      type: "string",
    },
    ownshares: {
      type: "string",
    },
    ciscode: {
      type: "string",
    },
    bsrcode: {
      type: "string",
    },
    atmfacilityrequired: {
      type: "string",
    },
    atmcardstatus: {
      type: "string",
    },
    smsrequired: {
      type: "string",
    },
    stmtfrequency: {
      type: "string",
    },
    incomerecptmode: {
      type: "string",
    },
    lineofbusiness: {
      type: "string",
    },
    politicallyexposed: {
      type: "string",
    },
    countryofrisk: {
      type: "string",
    },
    crossborderrisk: {
      type: "string",
    },
    domesticrisk: {
      type: "string",
    },
    risktype: {
      type: "string",
    },
    segmentcode: {
      type: "string",
    },
    creditthreshold: {
      type: "string",
    },
    debitthreshold: {
      type: "string",
    },
    domicile: {
      type: "string",
    },
    ekycconsentflag: {
      type: "string",
    },
    authcode: {
      type: "string",
    },
    ddate: {
      type: "string",
    },
    dateandtime: {
      type: "string",
    },
    gatewayUrl: { type: "string", minLength: 1 },
    gatewayTimeout: { type: "number" },
    gatewayCaCertificate: { type: "string" },
    key: { type: "string" },
    cert: { type: "string" },
  },
  required: [
    "aofno",
    "enrollmenttype",
    "accsubtype",
    "branch",
    "custtittle",
    "custname",
    "custmiddlename",
    "custlastname",
    "custdob",
    "custage",
    "custsex",
    "cusrmartialstatus",
    "mothertitle",
    "motherfirstname",
    "motherlastname",
    "custcommunitycode",
    "custnationality",
    "mobilecountrycode",
    "custaadhaarno",
    "custmobilenumber",
    "custemailid",
    "custplaceofbirth",
    "educationalqualification",
    "presentaddresstype",
    "custaddress1",
    "custaddress2",
    "custaddress3",
    "custpresentlandmark",
    "custpresentlocality",
    "custvillage",
    "custcitycode",
    "custsubdist",
    "custdistrict",
    "custstatecode",
    "custcountry",
    "custpinnumber",
    "custpresentpostoffice",
    "permaddresstype",
    "custpermaddress1",
    "custpermaddress2",
    "custpermaddress3",
    "custpermlandmark",
    "custpermlocality",
    "custpermvillage",
    "custpermcitycode",
    "custpermsubdist",
    "custpermdistrict",
    "custpermstatecode",
    "permcountry",
    "custpermpincode",
    "custpermpostoffce",
    "custresidenttype",
    "custresidenceperiod",
    "custnetworth",
    "custocupationcode",
    "customerevaluationrequired",
    "custpancardpresense",
    "custtype",
    "custhousetype",
    "custemployertype",
    "custincomesource",
    "custtaxslab",
    "custform60or61",
    "custformpancardnumber",
    "custvehicletype",
    "custlanguagecode",
    "custannualexpectedcredit",
    "grossannualincome",
    "sourceannualincome",
    "addressproofdocnumber",
    "addressproofdoctype",
    "addrissuedate",
    "addrissuedby",
    "addrissueplace",
    "identitydocnumber",
    "identitydoctype",
    "idexpirydate",
    "idissuedate",
    "idissuedby",
    "idissueplace",
    "relationshiptype",
    "relationshipname",
    "relationshiptitle",
    "relfirstname",
    "rellastname",
    "nomineetitlecode",
    "nomineename",
    "nomineelastname",
    "nomineedob",
    "nomineeage",
    "nomineeaddress",
    "nomineerelation",
    "nomineetelephone",
    "nomineeminor",
    "guardianname",
    "guardianaddress1",
    "housingloan",
    "carloan",
    "educationalloan",
    "consumerloan",
    "businessagricultureloan",
    "ownbankfixeddeposit",
    "owngold",
    "ownmutualfund",
    "ownnsckvc",
    "ownppf",
    "ownproperty",
    "ownrbibond",
    "ownshares",
    "ciscode",
    "bsrcode",
    "atmfacilityrequired",
    "atmcardstatus",
    "smsrequired",
    "stmtfrequency",
    "incomerecptmode",
    "lineofbusiness",
    "politicallyexposed",
    "countryofrisk",
    "crossborderrisk",
    "domesticrisk",
    "risktype",
    "segmentcode",
    "creditthreshold",
    "debitthreshold",
    "domicile",
    "ekycconsentflag",
    "authcode",
    "date",
    "dateandtime",
    "gatewayUrl",
    "gatewayTimeout",
  ],
};

const SIGNUploadSchema = {
  type: "object",
  title: "SIGNATURE UPLOAD",
  properties: {
    customerId: {
      type: "string",
    },
    accountNo: {
      type: "string",
    },
    encodedSignature: {
      type: "string",
    },
    gatewayUrl: { type: "string", minLength: 1 },
    gatewayTimeout: { type: "number" },
    gatewayCaCertificate: { type: "string" },
    key: { type: "string" },
    cert: { type: "string" },
  },
  required: [
    "customerId",
    "accountNo",
    "encodedSignature",
    "gatewayUrl",
    "gatewayTimeout",
  ],
};

module.exports = {
  ACCTandCIFCreationSchema,
  SIGNUploadSchema,
};
